content/flexible-types.md

```diff
@@ -0,0 +1,179 @@
+
+This is not a SIP for explicit nulls itself, but only for standardizing the representation of flexible types in the TASTy format.
+
+## Motivation
+
+### Background: Explicit Nulls and Java/Legacy Scala Interoperability
+
+When explicit nulls are enabled (`-Yexplicit-nulls`), Scala's type system changes so that `Null` is no longer a subtype of reference types. Instead, nullable types must be explicitly declared as union types like `String | Null`. This creates a safer type system but introduces friction when interoperating with:
+
+1. **Java libraries**, where all reference types are implicitly nullable
+2. **Legacy Scala code**, compiled without explicit nulls, where reference types could historically contain `null` values
+
+### The Problem
+
+Consider a Java method with signature `String getName()` or a legacy Scala method `def getName(): String` compiled without explicit nulls. Under explicit nulls,
+
+1. **If we type it as `String`**: We lose safety because the Java method or legacy Scala method might actually return `null`
+2. **If we type it as `String | Null`**: We burden users with constant null checks even when the method is known to never return null in practice
+
+### Current Workarounds and Their Limitations
+
+Before Flexible Types, the compiler would either:
+- Force all Java and legacy Scala reference types to be nullable (`String | Null`), leading to excessive null-checking
+- Provide unsafe nulls mode (`-language:unsafeNulls`) which disables safety checks entirely
+
+Both approaches are suboptimal for large codebases that want gradual migration to explicit nulls.
+
+## Proposed solution
+
+### High-level overview
+
+We introduce Flexible Types as an Internal Type that allows a type to be treated as both nullable and non-nullable depending on the context. Informally, we write a flexible type as `T?` (notation inspired by Kotlin platform types). The following subtyping relationships hold: `T | Null <: T?` and `T? <: T`, meaning:
+
+- It can accept both `T` and `T | Null` values
+- It can be used where either `T` or `T | Null` is expected
+- It can be used as the prefix in accesses to members of `T`, but may throw `NullPointerException` at runtime if the value is actually null
+
+Flexible Types are **non-denotable** - users cannot write them explicitly in source code. Only the compiler creates them during Java interoperability and when consuming legacy Scala code.
+
+They may appear in type signatures because of type inference.
+Due to their non-denotable nature, we do not recommend exposing Flexible Types in public APIs or library interfaces.
+We have implemented a mechanism to warn users when Flexible Types are exposed at public (or protected) field or method boundaries.
```

Comment:
I think it'd be good if the SIP could describe in more detail this warning mechanism and other situations where users might be confronted with flexible types. Ideally with a comparison to how platform types are handled in Kotlin since they have a similar nature and I assume have had time to get a polished user experience.

Reply:
(or maybe this comes up in the follow-up explicit nulls SIP, not sure where to draw the line)